### PR TITLE
Explain blocking versus non-blocking changes in PRs

### PR DESCRIPTION
**Pull request by staltz**
See https://github.com/ssb-js/ssb-js/pull/18#issuecomment-687035278 for more context.

The idea is to avoid debatable terms such as "small improvements" or "net positive" and instead require maintainers to communicate (before PR creation) what blocks a PR from merging (is it tests? is it security concerns? is it a large PR that conflates many changes?), and require contributors to update their PRs until they don't have blocks.

Note that this requires projects to have "CONTRIBUTING.md" files, which I anyway think we should have.

### CODE REVIEW

File: README.md
Comment (christianbundy) [inline README.md:79]: :+1: I think this captures the spirit of the previous rules very nicely.

File: README.md
Comment (christianbundy) [inline README.md]: Clarification: Could you add some details on what this might look like? I'm having a hard time imagining writing down all the things that might block a PR, and I'm wondering whether this might already be covered in the next point: "communicate [...] what is expected in a pull request".

File: README.md
Comment (christianbundy) [inline README.md]: I have a small concern around this requirement -- more info (and a patch) here: https://github.com/ssb-js/ssb-js/pull/23

**Review by christianbundy — COMMENTED** [review]
This looks good. I'd like a small clarification on the `CONTRIBUTING.md` and have a small concern over the new 'Contributors MUST' language, but the meaty part of this PR looks great to me:

>  Maintainers SHOULD NOT suggest non-blocking changes on project pull requests. 

- 8791118: Huge :+1:, I would approve this if it were alone.
- 42c6eee: Some concerns, would like to discuss.

File: README.md
Comment (staltz) [reply README.md]: That's interesting! If you feel like many things could block a PR, that's a good sign we should write them down explicitly as opposed to keeping it implicit.

The "things that are expected in a PR" may not suffice to define blocking concerns. E.g. what if the PR contains 400 commits, only 1 of those commits fixes the issue at hand. At the same time, code style is changed everywhere, node_modules is committed into git, and the LICENSE is deleted. The thing we expected, the fix, **is** there, but so are a bunch of unexpected things.

In practice, contributing.md could list things like:

"We won't accept PRs that do too many changes at once, or do side effects unrelated to the issue at hand. We won't accept PRs that create new security vulnerabilities." Etc

**Comment by staltz** [comment]
Thanks for the suggestions. I could also make a new PR, but let's see first what other maintainers comment. And right now I'm on mobile, can't do edits.

File: README.md
Comment (christianbundy) [reply README.md]: > The thing we expected, the fix, is there, but so are a bunch of unexpected things.

C4 solves this problem by identifying a patch as "a minimal and accurate answer to exactly one identified and agreed problem". I think we should write down the rules we frequently enforce, especially if we can write them as tests, but we should also give maintainers the ability to say "we haven't written rules about this yet, but I'm requesting changes".

The number of potential problems is infinite, and I think we'd be better off with:

- A general expectation of what patches should look like.
- Tests that a user can run locally to ensure that their patch conforms.
- The ability for maintainers to say "please change X" without requiring that it's already written in the contributor guide.

In other words: I'm worried that this rule forces maintainers to merge anything that doesn't break an already-written-down rule.

```suggestion
- Maintainers SHOULD communicate in the project's CONTRIBUTING.md what types of changes in a project pull request are frequently blocked from merging.
```

File: README.md
Comment (soapdog) [reply README.md:79]: 👍 I like this too.

**Comment by soapdog** [comment]
> Thanks for the suggestions. I could also make a new PR, but let's see first what other maintainers comment. And right now I'm on mobile, can't do edits.

I like this PR and it is OK to merge for me provided we internalize that some PRs will be closed without merging because the contributor will not want to go through with the requested changes (that is OK for me).

File: README.md
Comment (staltz) [reply README.md]: Resolved

**Comment by staltz** [comment]
Ok friends, I took #23 and #18 and all your comments into consideration and updated this PR. I recommend reading through all the changes and checking whether y'all agree.

File: README.md
Comment (staltz) [reply README.md:79]: :+1:

**Review by arj03 — APPROVED** [review]

**Review by christianbundy — APPROVED** [review]
Love it, can't wait to merge!

**Review by soapdog — APPROVED** [review]

**Comment by christianbundy** [comment]
That's a bingo.

**Pull request merged**
